### ISSUE TITLE
/admin/login/?next=/admin/ won't redirect when logging in with microsoft credentials

### ISSUE THREAD

**Issue by greglever** [AngellusMortis/django_microsoft_auth#129]
When I navigate to /admin/login/?next=/admin/ and log in with 
an existing username and password I get logged in and redirected to the admin panel.

However, when I log in with my microsoft credentials, I can successfully create and log in a user but I don't get redirected, it just stays on the same page.

I know the user has been created and logged in correctly because if I then navigate to /admin I am logged in as that user.

Any ideas how I can get the redirect to work ?

**Comment by AngellusMortis** [comment]
Have you check your console for Javescript errors? All of that is done via Javascript. Also, what browser, browser version and OS are you using?

________________________________

**Comment by greglever** [comment]
so when I log in with exsting admin credentials I see:
```
Starting development server at http://0.0.0.0:8000/
Quit the server with CONTROL-C.
[02/Nov/2018 15:57:13] "GET /admin/login/?next=/admin/ HTTP/1.1" 200 3313
[02/Nov/2018 15:57:13] "GET /static/microsoft/css/admin_login.css HTTP/1.1" 200 201
[02/Nov/2018 15:57:13] "GET /static/microsoft/js/admin_login.js HTTP/1.1" 200 4322
[02/Nov/2018 15:57:26] "POST /admin/login/?next=/admin/ HTTP/1.1" 200 3500
[02/Nov/2018 15:57:26] "GET /static/microsoft/css/admin_login.css HTTP/1.1" 304 0
[02/Nov/2018 15:57:26] "GET /static/microsoft/js/admin_login.js HTTP/1.1" 304 0
[02/Nov/2018 15:57:33] "POST /admin/login/?next=/admin/ HTTP/1.1" 302 0
[02/Nov/2018 15:57:33] "GET /admin/ HTTP/1.1" 200 5601
```
and then when I log in via microsoft credentials I see:
```
[02/Nov/2018 15:58:13] "GET /admin/logout/ HTTP/1.1" 200 1207
[02/Nov/2018 15:58:20] "GET /admin/ HTTP/1.1" 302 0
[02/Nov/2018 15:58:20] "GET /admin/login/?next=/admin/ HTTP/1.1" 200 3313
[02/Nov/2018 15:58:20] "GET /static/microsoft/css/admin_login.css HTTP/1.1" 200 201
[02/Nov/2018 15:58:20] "GET /static/microsoft/js/admin_login.js HTTP/1.1" 200 4322
[02/Nov/2018 15:58:22] "POST /microsoft/auth-callback/ HTTP/1.1" 200 404
```
The value of `context` here:
https://github.com/AngellusMortis/django_microsoft_auth/blob/25ac0195ac7a052034359211b70c770c905b549f/microsoft_auth/views.py#L110
is 
```
{'base_url': 'http://0.0.0.0:8000/', 'message': '{}'}
```

So I'm assuming the 404 in the last:
```
[02/Nov/2018 15:58:22] "POST /microsoft/auth-callback/ HTTP/1.1" 200 404
```
line is something to do with http://0.0.0.0:8000/ not existing ?

This is on Ubuntu 18.04 LTS on chrome

**Comment by AngellusMortis** [comment]
> line is something to do with http://0.0.0.0:8000/ not existing ?

Yes. You *must* set up and configure the Sites Framework correctly for the domain you will be access the site from. If you are going to be using http://localhost:8000 to access the site, you *must* have set up in the admin as your current domain. `AuthenticateCallbackView` uses the value configured in via your Sites Framework to prevent host headers attacks (the `HOST` value inside of the `request` object is not considered "safe" since it comes from the HTTP Host header. 

The redirect is done via posting messages between browser windows. Do the domains *must* match up perfectly and preferably be HTTPS. Browsers can be picky about posting messages across windows if they are not both secure. If you checkout the [template for `auth-callback`](https://github.com/AngellusMortis/django_microsoft_auth/blob/25ac0195ac7a052034359211b70c770c905b549f/microsoft_auth/templates/microsoft/auth_callback.html), you can see it is just a `DOMContentLoaded` event that posts a message to the parent window (this would be the `/admin/login` page) the result of the login attempt.

**Comment by greglever** [comment]
OK thanks, yeah I've got the Sites Framework configured. I've put a swagger on http://localhost:8000/ but still getting that last 404, it must be to do with it being http then instead of https.

**Comment by AngellusMortis** [comment]
Yeah, 'https' is hardcoded on the `auth-callback` URL. I am pretty sure `localhost` might be a whitelisted HTTP URL from Microsoft. 

I mostly develop from a Windows machine, so I am often using a VM or a remote dev server that is Linux to actually run stuff, so I often not use `localhost` which is why I think I thought HTTPS was a main requirement. As a result, I did not really put a lot of effort into trying to figure out if the page request was made with HTTP vs. HTTPS on the backend. 

This weekend I can see if I can try to get `localhost` to work over HTTP and improve the error messaging around the Sites Framework and such.

**Comment by greglever** [comment]
> In a production environment though, you should obviously always be using HTTPS though.

Agreed, definitely

> Yeah, 'https' is hardcoded on the auth-callback URL. I am pretty sure localhost might be a whitelisted HTTP URL from Microsoft.

Initially I was working on my fork of this repo and manually changed them to http from https, but I've changed them back now

I'm currently using [localtunnel](https://localtunnel.github.io/www/) to expose an https url and route it to my local server but strangely I still get messages saying:
```
Security Warning 

The information you have entered on this page will be sent over an insecure connection and could be read by a third party.

Are you sure you want to send this information?
```
in firefox but nothing in chrome. In any case I still can't get the redirect to work...

**Comment by AngellusMortis** [comment]
Sorry it took so long this update everything, but this should be resolved now in `1.2.0`. There was Javascript to pass messages between the two login Windows (Django Admin login and Microsoft login) and this was not functioning currently if your URL was on any port other than 80 for HTTP or 443 for HTTPS. 

I also added support for `http://localhost`. HTTP authentication only works for `localhost` though and only if `DEBUG` is enabled. 

Can you try to verify any more issues on your end? Otherwise I will close out the ticket in 7 days.
